### PR TITLE
Add Right To Know stats, and put runners in one place

### PR DESCRIPTION
**Pull request by auxesis**
Looks like this: 

image

This effectively deprecates the need for [openaustralia/jacaranda-righttoknow](https://github.com/openaustralia/jacaranda-righttoknow).

Also: 

 - Split the scraper out into runners under `lib/runners`, to reduce cognitive overhead for maintainers. 
 - Make it easy to add new runners to post stats for different applications (based [on feedback](https://github.com/openaustralia/jacaranda/pull/5/files#r133626629) from @equivalentideas).
 - Improve the test suite coverage by testing behaviour after we time travel.

The intention is to make Jacaranda easier to understand and contribute to.

### CODE REVIEW

**Comment by equivalentideas** [comment]
@auxesis for me this is clearer 👍 

There's some other feedback on smaller stuff on the other PR that would be good to bring over, and does the readme have to update for this version?

I'm also interested in why we're adding the command line options stuff to this, rather than have it controlled through environment variables for morph.

I know you're a command line wizard ;D but the need isn't apparent to me. Have you got a specific use case in mind?

If you're happy with this and have looked at the feedback on the other PR I'm happy for you to merge this 👍 don't wanna hold it up.

**Comment by auxesis** [comment]
Feedback from #5 that needs to be checked and incorporated: 

- [x] https://github.com/openaustralia/jacaranda/pull/5/files#r133619845
- [x] https://github.com/openaustralia/jacaranda/pull/5/files#r133624538
- [x] https://github.com/openaustralia/jacaranda/pull/5/files#r133620585
- [x] https://github.com/openaustralia/jacaranda/pull/5/files#r133620683
- [x] https://github.com/openaustralia/jacaranda/pull/5/files#r133624637
- [x] https://github.com/openaustralia/jacaranda/pull/5/files#r133623744
- [x] https://github.com/openaustralia/jacaranda/pull/5/files#r133622235
- [x] https://github.com/openaustralia/jacaranda/pull/5/files#r133622424
- [x] Add a `MORPH_RUNNERS` lookup to set the runners filter.
- [x] Update README for the new way of adding runners.

**Comment by auxesis** [comment]
@equivalentideas 

Have [added a line to the README](https://github.com/openaustralia/jacaranda/pull/8/commits/8f187b20608c5b3ea2b11246adca60301ccc8a79) explaining why the `--runners` option was added. 

The only reason it's there is to aid debugging when contributors are adding new runners. I definitely wouldn't expect it to be used this way when it's run on Morph. You're right though – we could change this to be an environment variable, given we're already gone all in on dotenv. 

I'll update the rest of the README to match the new way of adding runners.

**Comment by equivalentideas** [comment]
> The only reason it's there is to aid debugging when contributors are adding new runners. I definitely wouldn't expect it to be used this way when it's run on Morph. You're right though – we could change this to be an environment variable, given we're already gone all in on dotenv.

That sounds reasonable to me @auxesis 👍 thanks as always 💓

**Comment by auxesis** [comment]
@equivalentideas this is all fixed up now. 

Let me know if you're good with the last few documentation changes, and I'll merge.

File: spec/runner_spec.rb [hunk 1]
```diff
@@ -159,10 +159,11 @@ def all_request_bodies
   end
 
   describe '.runners' do
-    context 'when filtering' do
+    context 'when filtering with cli option' do
       it 'returns everything by default' do
         Jacaranda.parse([])
-        expect(Jacaranda.runners.size).to be >= mock_runner_count
+        runners = Jacaranda.runners & mock_runners
```
Comment (equivalentideas) [inline spec/runner_spec.rb]: I do find these tests a bit confusing. It seems like `.parse` is effectively being tested throughout here as well?

It's confusing that we're testing `.runners` but then using the [`&` method](http://ruby-doc.org/core-2.4.1/Array.html#method-i-26) to pick out the common elements in these two arrays, manipulating the result of the execution before we verify them.

**Review by equivalentideas — CHANGES_REQUESTED** [review]
@auxesis I think this is a better structure 👍 

[I've commented](https://github.com/openaustralia/jacaranda/pull/8/files#r136479898) on the tests for `Jacaranda.runners` because I find them quite confusing—it seems like there's some prerequisite knowledge that isn't in the test that you need to understand what the expected behavior is. If you think it's feasible, I'd like to improve those before merging this.

Aside from that, this looks good to go ✨

File: spec/runner_spec.rb [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+        restore_env
+      end
+    end
+  end
+
+  describe '.runners' do
+    context 'when filtering with cli option' do
+      it 'returns everything by default' do
+        Jacaranda.parse([])
+        runners = Jacaranda.runners & mock_runners
```
Comment (equivalentideas) [inline spec/runner_spec.rb]: I do find these tests a bit confusing. It seems like `.parse` is effectively being tested throughout here as well?

It's confusing that we're testing .runners but then using the `&` method to pick out the common elements in these two arrays, manipulating the result of the execution before we verify them—which makes it unclear what `.runners` is actually expected to do.

Comment (auxesis) [reply spec/runner_spec.rb]: Yeah, so the `&` is to work around a nasty bug in the tests. 

The tests create mock runner classes, but don't seem to destroy them after each scenario. 

I'll do some more digging to see if I can fix this.

Comment (equivalentideas) [reply spec/runner_spec.rb]: > Yeah, so the & is to work around a nasty bug in the tests.

Cool, that definitely sounds like something we want to fix before merging 👍 Feel free to bring me into the nitty gritty if it helps :)

**Comment by auxesis** [comment]
I've had to do some fairly significant refactoring of the tests, and things are broken again. 

Will come back to this when I have time.

**Comment by auxesis** [comment]
@equivalentideas have refactored the tests per your feedback. Ready for another review.

**Review by equivalentideas — APPROVED** [review]

**Comment by equivalentideas** [comment]
I'm merged this @auxesis thanks so much for this huge contribution. I'll disable the Right To Know jacaranda on morph.io now and just let this one run.

**Comment by auxesis** [comment]
Argh. After all this work, I just realised that there's caveats listed in #5 that I haven't addressed in this PR:

> Caveats:
> 
> This PR changes the primary key from date_posted to a composite primary key of date_posted and runner.
> I need to do further testing to see how the scraper handles this when run on Morph (the testing behaviour locally was not great).

I'll do a quick test this evening and double check what happens. 

If I discover problems, I'll submit another PR.

**Comment by equivalentideas** [comment]
> Argh. After all this work, I just realised that there's caveats listed in #5 that I haven't addressed in this PR

All good @auxesis with such a big PR changing so much it's so easy to miss stuff like that.

**Comment by auxesis** [comment]
Forgot to update last night, but there are problems: #15.

**Pull request merged**